From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 21 Oct 2025 21:11:34 -0700
Subject: [PATCH] Update Paper repo URL


diff --git a/build.gradle.kts b/build.gradle.kts
index d722f44da6c1114459daedb43da804c444009fa9..064b190cc95795360453f4e79777be0086c01ab5 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -46,7 +46,7 @@ subprojects {
         maven("https://repo.codemc.io/repository/maven-public")
         maven("https://oss.sonatype.org/content/repositories/snapshots/")
         maven("https://s01.oss.sonatype.org/content/repositories/snapshots/")
-        maven("https://papermc.io/repo/repository/maven-public/")
+        maven("https://repo.papermc.io/repository/maven-public/") // Insights-Patched - Update Paper repo URL
         maven("https://libraries.minecraft.net")
     }
 
diff --git a/settings.gradle.kts b/settings.gradle.kts
index c36a94ff600273b8510460141d99aebc11041efd..96c09f4f752917ad21cf70f3c5e25289808eef21 100644
--- a/settings.gradle.kts
+++ b/settings.gradle.kts
@@ -3,7 +3,7 @@ import java.nio.file.Files
 pluginManagement {
     repositories {
         gradlePluginPortal()
-        maven("https://papermc.io/repo/repository/maven-public/")
+        maven("https://repo.papermc.io/repository/maven-public/") // Insights-Patched - Update Paper repo URL
     }
 }
 
