From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 21 Oct 2025 21:41:15 -0700
Subject: [PATCH] Always disable the update checker


diff --git a/Insights-Core/src/main/java/dev/frankheijden/insights/Insights.java b/Insights-Core/src/main/java/dev/frankheijden/insights/Insights.java
index cfe67e77f6b009321894f2c0faf6a1aafbbc878e..742c42f68b3482c1b1d4821c82779f73af6e462d 100644
--- a/Insights-Core/src/main/java/dev/frankheijden/insights/Insights.java
+++ b/Insights-Core/src/main/java/dev/frankheijden/insights/Insights.java
@@ -391,7 +391,7 @@ public class Insights extends InsightsPlugin {
             updateChecker.cancel();
         }
 
-        if (settings.UPDATE_CHECKER_ENABLED) {
+        if (false) { // Insights-Patched - Always disable the update checker
             updateChecker = getServer().getScheduler().runTaskTimerAsynchronously(
                     this,
                     new UpdateCheckerTask(this),
