From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 21 Oct 2025 22:36:29 -0700
Subject: [PATCH] Remove checkstyle


diff --git a/build.gradle.kts b/build.gradle.kts
index 1660726d338edafc3bc1041ce832c1030a37622a..5770400a404c1664981ff7c6c829f16578badbad 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -19,7 +19,6 @@ val targetMinecraftVersions = listOf("26.1.2", "1.21.11", "1.21.10", "1.21.9")
 
 subprojects {
     apply(plugin = "java")
-    apply(plugin = "checkstyle")
     apply(plugin = "com.gradleup.shadow")
 
     version = rootProject.version
@@ -77,7 +76,7 @@ subprojects {
 
     tasks {
         build {
-            dependsOn("shadowJar", "checkstyleMain", "checkstyleTest", "test")
+            dependsOn("shadowJar", "test") // Insights-Patched - Remove checkstyle
         }
 
         compileJava {
@@ -100,13 +99,6 @@ subprojects {
         }
     }
 
-    tasks.withType<Checkstyle>().configureEach {
-        configFile = file("$rootDir/config/checkstyle/checkstyle.xml")
-        ignoreFailures = false
-        maxErrors = 0
-        maxWarnings = 10
-    }
-
     tasks.withType<ShadowJar> {
         relocate("dev.frankheijden.minecraftreflection", "$dependencyDir.minecraftreflection")
         relocate("io.papermc.lib", "$dependencyDir.paperlib")
