From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 22 Oct 2025 00:36:05 -0700
Subject: [PATCH] Add fork identifiers to plugin.yml


diff --git a/Insights-Core/build.gradle.kts b/Insights-Core/build.gradle.kts
index c1c5f0b8cecd2574fcab301bad0b511bbbeb88bb..14937634133fd25a4bb0f87f1eafae6a58940e89 100644
--- a/Insights-Core/build.gradle.kts
+++ b/Insights-Core/build.gradle.kts
@@ -34,7 +34,10 @@ bukkit {
     main = "dev.frankheijden.insights.Insights"
     description = "Insights about your server and regional block limits"
     apiVersion = "1.21"
-    website = "https://github.com/InsightsPlugin/Insights"
+    // Insights-Patched start - Add fork identifiers to plugin.yml
+    version = "${project.version}-AcheronInteractive"
+    website = "https://github.com/AcheronInteractive/Insights"
+    // Insights-Patched end - Add fork identifiers to plugin.yml
     softDepend = listOf("PlaceholderAPI")
     authors = listOf("FrankHeijden", "Jsinco")
     permissions {
