From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 21 Oct 2025 22:36:29 -0700
Subject: [PATCH] Remove checkstyle


diff --git a/build.gradle.kts b/build.gradle.kts
index 67ca77bd8eb84b5b4b597470f9b693f1bfae8834..1ba89523d033871da6793de065760b2fc36fa2ba 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -16,7 +16,6 @@ version = "6.20.0" // Insights-Patched - Custom version
 
 subprojects {
     apply(plugin = "java")
-    apply(plugin = "checkstyle")
     apply(plugin = "com.gradleup.shadow")
 
     version = rootProject.version
@@ -74,7 +73,7 @@ subprojects {
 
     tasks {
         build {
-            dependsOn("shadowJar", "checkstyleMain", "checkstyleTest", "test")
+            dependsOn("shadowJar", "test") // Insights-Patched - Remove checkstyle
         }
 
         compileJava {
@@ -97,13 +96,6 @@ subprojects {
         }
     }
 
-    tasks.withType<Checkstyle>().configureEach {
-        configFile = file("$rootDir/config/checkstyle/checkstyle.xml")
-        ignoreFailures = false
-        maxErrors = 0
-        maxWarnings = 0
-    }
-
     tasks.withType<ShadowJar> {
         relocate("dev.frankheijden.minecraftreflection", "$dependencyDir.minecraftreflection")
         relocate("io.papermc.lib", "$dependencyDir.paperlib")
