From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 22 Oct 2025 01:26:01 -0700
Subject: [PATCH] Fix error trying to display air as an item


diff --git a/Insights-API/src/main/java/dev/frankheijden/insights/api/config/Messages.java b/Insights-API/src/main/java/dev/frankheijden/insights/api/config/Messages.java
index 9a4a719243caf5892b583d6601c1ed5f2557aff4..798be5971c5e5592ecdf24ac189544029c7dea2d 100644
--- a/Insights-API/src/main/java/dev/frankheijden/insights/api/config/Messages.java
+++ b/Insights-API/src/main/java/dev/frankheijden/insights/api/config/Messages.java
@@ -86,7 +86,7 @@ public class Messages {
 
             if (obj instanceof Material material) {
                 var key = material.getKey();
-                if (material.isItem()) {
+                if (material.isItem() && !material.isAir()) { // Insights-Patched - Fix error trying to display air as an item
                     return component.hoverEvent(HoverEvent.showItem(
                             net.kyori.adventure.key.Key.key(key.getNamespace(), key.getKey()),
                             1
