From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Wed, 22 Oct 2025 04:03:23 -0700
Subject: [PATCH] Mark plugin as using mojang mappings


diff --git a/build.gradle.kts b/build.gradle.kts
index 989493f1d88667e33c99effeab21372696f54787..4ac44ce14cf9dcafa1679d03caddb35da76baf90 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -124,6 +124,7 @@ tasks {
         dependsOn("shadowJar")
         finalizedBy("copyJars")
     }
+    jar { manifest { attributes["paperweight-mappings-namespace"] = "mojang" } } // Insights-Patched - Mark plugin as using mojang mappings.
 }
 
 tasks.register("cleanJars") {
